### DISCUSSION TITLE
Flash strategy for one plus

### DISCUSSION THREAD

**Discussion by freechelmi** [openandroidinstaller-dev/openandroidinstaller#377] [General]
I'm not sure how to resolve problems we have with one plus devices here 

https://github.com/openandroidinstaller-dev/openandroidinstaller/issues/298

Basically they cannot boot the recovery 

Can you first detail the current steps and why  : 

- Oem unlock
- Boot recovery temp
- Flash OS 
- Flash dtbo & vbmeta 
- Flash recovery ?

**Comment by tsterbak** [comment]
Hey, thanks for starting this discussion! It would be great to resolve these issues.
Unfortunately, I have one OP devices available at the moment. Maybe I can get my hands on some until then I have to rely on other people to debug and test it.

The current steps of the OAI for the OnePlus 7T Pro (hotdog) are: 
- oem unlock
- flash dtbo & vbmeta with fastboot
- boot twrp with fastboot temporarily (decided since it is a A/B device)
- sideload OS
- reboot

I notice now, that the LOS instructions want to flash the recovery here. So maybe we should do that as well. I thought its risky to flash TWRP directly on A/B partitioned devices, but for OP I might be wrong.

How would you do the process manually with TWRP? Does flashing TWRP from fastboot work directly?

**Reply by freechelmi** [reply]
I could send you a one plus A/B susch as the 8 pro if I have one soon , let me know your postal adress on Telegram  https://t.me/freechelmi

**Reply by freechelmi** [reply]
I'm kind of block because even manually i could not flash the TWRP twrp-3.7.0_11-0-kebab.img although I've flashed on sell several before , I need to dig this tomorrow

**Comment by freechelmi** [comment]
After few tries : 

- I'm stuck even flashing manually the One plus 8T ( Kebab KB2003 ) which does not seems different from before , except it came with OXYOS Android 13 instead of android 12 in october as I recall
- I've tried to add --disable-verity  when flashing vbmeta and dtbo even if it sounds dumb as the bootloader is already unlock 

Any help appreciated :-) 

![IMG_20231204_143025 resized resized](https://github.com/openandroidinstaller-dev/openandroidinstaller/assets/446021/babd62ae-743c-4df4-bb28-e445e3488910)

**Comment by freechelmi** [comment]
I've also tried to boot the boot.img from lineageos without success 

I'm missing a piece of info it seems 
fastboot boot kebab/lineageos/boot.img

**Comment by freechelmi** [comment]
Some progress : 
- No version version of TWRP could be booted ( like 6 months ago actually ) 
- I could flash/boot the eOS recovery ( but this does not work for OAI ....)

**Reply by tsterbak** [reply]
Maybe I should work on a version of the Installer that works with the eOS or official LOS recovery... I'm definitely thinking about it!

**Reply by freechelmi** [reply]
Until LOS recovery integrate some commands like twrp ( not the case right now AFAIK ) we won't be able to use it I think

**Reply by tsterbak** [reply]
Yeah, it would not be as comfortable as it is with TWRP. But at least all manual steps in the LOS recovery look the same (as far as I know), so we could have default instructions on how to do it. Very early versions of the installer included this (very early work on this can be found in this script for example: https://github.com/openandroidinstaller-dev/openandroidinstaller/blob/main/scripts/lineageos-on-pixel3a.py)

But I agree, it would be quite some work to include it properly...

**Comment by freechelmi** [comment]
OK almost working with this unofficial orangefox version https://bit.ly/3T9idxc

Last operation was adb shell twrp wipe dalvik when orangefox rebooted the device in an unbootable state 

After forcing a wipe data , eOS booted

**Reply by tsterbak** [reply]
Getting Ofox to work would be great, since support for it might be added soon.
Did you make any new progress on this?

**Comment by freechelmi** [comment]
OK il missed a really Big problem with OAI  ( 0.5.1 )  , example for OnePlus 5T

- The TWRP recovery is actually never written , it's just booted once 
- Then when an update comes ( eOS here) , the update cannot be installed by the oneplus OEM recovery 
- I had the problem with few client and I don't think I'll be able to fix the problem remotely

**Reply by ghost** [reply]
Uh, it's due to `fastboot boot ...` instead of `fastboot flash recovery ...` (my xiaomi configs aren't affected my this issue)
Remotely, the only way I think to do that is to remote control the computer, that's not handy

**Reply by freechelmi** [reply]
Yes but I thought after the boot recovery , we write it , I was wrong

**Reply by tsterbak** [reply]
I'm sorry for the annoyance :/

Currently (especially for A/B partitioned devices, other than samsung and xiaomi), the installer only boots TWRP (as you noticed). This is done for two reasons:
1. I only do clean flashes and no updates (which is probably an antipattern at this point...)
2. Flashing TWRP to a/b partitioned devices can be dangerous and if something goes wrong much harder to recovery, while, when no recovery is flashed, no harm has been done.

We can move forward and try to change the behavior :) What would be the expected behavior in our opinion?

**Reply by freechelmi** [reply]
You don't need to apologize ! You already made such a difference with OAI , combined with project like eOS , we are pushing smartphones to the next level ! :-)

One point is that it does not seems that "is_ab" determine of the recovery will be booted or flashed , I'm digging it right to at least implement this expected behaviour

**Reply by freechelmi** [reply]
So if I take dumpling OP5T as example : 
- The Flash or Boot recovery is defined in the Device config steps
- Our main problem is that we cannot reboot to recovery from fastboot ( or download mode anyway)  
- Today we ( in 0.5.1) After the Boot unlock 
-- Boot TWRP via fastboot
-- format data wi TWRP 
-- Wipe cache & system
-- Activate sideload
-- adb sideload rom.zip 
-- We then adb reboot , rom works but still have old recovery


I propose instead of rebooting at last step : 

-- adb reboot bootloader
-- fastboot  flash recovery twrp.img
-- fastboot reboot 

And then all is well , I will send a PR to dev for dumpling about this but let me know how you would want this to be arranged

**Reply by freechelmi** [reply]
Here we go https://github.com/openandroidinstaller-dev/openandroidinstaller/pull/406

**Reply by tsterbak** [reply]
I already commented on the PR, but maybe its better to have the discussion here :D

So if we know it's save to flash TWRP instead of booting it, by not flash it directly in the first place?
That should actually be true for the OP5T (checking here: https://twrp.me/oneplus/oneplus5-5t.html). So the `fastboot_boot_recovery` could be replaced by `fastboot_flash_recovery` in the config for this device, I guess.

**Reply by freechelmi** [reply]
AFAIK adb_reboot_recovery won't work on OP , it would just reboot to system and thus overwrite the recovery 

Or it would need the user to start manually on recovery which I would like to avoid as much as possible.

I guess it's time to create "families of devices" like OP OP_ab Xia Xia_ab to easily maintain all devices that behaves the same

**Reply by tsterbak** [reply]
I guess in the case of OP5, just flashing the recovery in the first place should be fine. @SirRGB correct me if I'm wrong

**Reply by SirRGB** [reply]
Yea, the op5 series has a recovery partition and is a-only. The latest oos 10.0.1 does work with both flashing and booting the latest official twrp, op3 series should be fine with that too.
I wouldnt want to flash the recovery after the rom install, because some devices ship a recovery (for ota purposes) with the rom package, which would be overridden by this. Ota can work fine with twrp, but isnt guaranteed, while ota on los recovery is a must for official releases.

**Comment by freechelmi** [comment]
So I propose a small videoconference meeting to discuss the right way to "modularize" the device config and attached flash strategy for the major brands.

What do you think about 3rd of January 4PM UTC ?

**Reply by tsterbak** [reply]
Hey, interesting idea :)

1. I don't really like calls, especially for this project. 
   - I like written communication and public discussion. This way everyone knows whats been discussed and can participate. Also it helps me remember things :D
   - I guess, most people participating here regularly are happy to stay pseudonymous and don't like participating in calls either.
2. I believe (and I might be wrong) that at this point, modularizing the configs for specific vendors is a form of premature abstraction that will make the project less flexible. Most devices even from the same vendor behave differently between generation (and even within).
This is why I prefer the more flexible way of non-specific configs. This way, the config can be tailored to every device specifically and there are less dependencies between devices. (For example, it looks like for the OP5T we could just directly flash twrp, while for other OP devices this might not be a good idea.) 

That said, I think we could move forward here by:
1. Check for specific vendors and devices (for example OP) if we could flash TWRP directly, adapt the config acordingly and then test it (or find someone to test it, or ship it as untested). This should be fine for quite some cases.
2. If we find cases where flashing twrp directly is not advisable, document in an issue what should be done for that device. If we have seen a few examples, maybe we can generalize from it.
     - I expect for example, that for some devices, we need to flash twrp from trwp after booting it. I need to figure out how to do this via scripting.
     
What do you think? :)

**Comment by freechelmi** [comment]
>     1. I don't really like calls, especially for this project.
I respect that

>     2. I believe (and I might be wrong) that at this point, modularizing the configs for specific vendors is a form of premature abstraction that will make the project less flexible. Most devices even from the same vendor behave differently between generation (and even within).

Agreed ,let's try to find the best behaviour for each model then

>        * I expect for example, that for some devices, we need to flash twrp from trwp after booting it. I need to figure out how to do this via scripting.

sound feasible ?
